components/merkledb/src/proof_list_index/proof.rs

```diff
@@ -184,67 +170,58 @@ fn merge(
     Merge::new(first, second)
 }
 
-/// Takes a subset of hashes at a particular height in the Merkle tree and
-/// computes all known hashes on the next height.
-///
-/// # Arguments
-///
-/// - `last_index` is the index of the last element in the Merkle tree on the given height.
-///
-/// # Return value
-///
-/// The `layer` is modified in place. An error is returned if the layer is malformed (e.g.,
-/// there is insufficient data to hash it).
-///
-/// # Examples
-///
-/// See unit tests at the end of this file.
-fn hash_layer(layer: &mut Vec<HashedEntry>, last_index: u64) -> Result<(), ListProofError> {
-    let new_len = (layer.len() + 1) / 2;
-    for i in 0..new_len {
-        let x = &layer[2 * i];
-        layer[i] = if let Some(y) = layer.get(2 * i + 1) {
-            // To be able to zip two hashes on the layer, they need to be adjacent to each other,
-            // and the first of them needs to have an even index.
-            if !x.key.is_left() || y.key.index() != x.key.index() + 1 {
-                return Err(ListProofError::MissingHash);
+fn hash_layer(layer: &[HashedEntry], last_index: u64) -> Result<Vec<HashedEntry>, ListProofError> {
```

Comment:
Seems that merging of `dynamic_services` into this branch was not quite successful.